### PR TITLE
Fail fast for integration tests

### PR DESCRIPTION
**Pull request by johnsonw**
- Integration testing should stop whenever there is a failure
- Add ability to create iml diagnostics with a provisioning script
- Add pub key to authorized keys for diagnostics upload
- Create diagnostic scripts in parallel on failure

Signed-off-by: johnsonw <wjohnson@whamcloud.com>

---
This change is [Reviewable](https://reviewable.io/reviews/whamcloud/integrated-manager-for-lustre/1827)

### CODE REVIEW

File: iml-system-docker-tests/tests/ldiskfs_test.rs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+async fn run_test(config: &vagrant::ClusterConfig) -> Result<(), SystemdError> {
+    run_fs_test(
+        &config,
+        &SetupConfigType::DockerSetup(SetupConfig {
+            use_stratagem: false,
+            branding: iml_wire_types::Branding::Whamcloud,
+        }),
+        vec![("base_monitored".into(), &config.storage_servers()[..])]
+            .into_iter()
+            .collect::<HashMap<String, &[&str], RandomState>>(),
```
Comment (jgrund) [inline iml-system-docker-tests/tests/ldiskfs_test.rs:21]: Why explicitly state the hasher?

File: iml-system-docker-tests/tests/ldiskfs_test.rs [hunk 2]
```diff
@@ -0,0 +1,38 @@
+// Copyright (c) 2020 DDN. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+mod utils;
```
Comment (jgrund) [inline iml-system-docker-tests/tests/ldiskfs_test.rs:5]: I don't think we need this

File: iml-system-docker-tests/tests/stratagem_test.rs [hunk 3]
```diff
@@ -0,0 +1,43 @@
+// Copyright (c) 2020 DDN. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+mod utils;
```
Comment (jgrund) [inline iml-system-docker-tests/tests/stratagem_test.rs:5]: I don't think we need this

File: iml-system-docker-tests/tests/ldiskfs_test.rs (hunk 1 above)
Comment (johnsonw) [reply iml-system-docker-tests/tests/ldiskfs_test.rs:21]: It was a clippy warning if I remember correctly. I can try taking it out again.

File: iml-system-docker-tests/tests/zfs_test.rs [hunk 4]
```diff
@@ -0,0 +1,39 @@
+// Copyright (c) 2020 DDN. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+mod utils;
```
Comment (jgrund) [inline iml-system-docker-tests/tests/zfs_test.rs:5]: I don't think we need this

File: iml-system-docker-tests/tests/ldiskfs_test.rs (hunk 2 above)
Comment (johnsonw) [reply iml-system-docker-tests/tests/ldiskfs_test.rs:5]: How do I access the utils functions if I take out the mod? I tried taking it out and it won't let me use them.

Comment (johnsonw) [reply iml-system-docker-tests/tests/ldiskfs_test.rs:5]: For visibility, I referenced https://doc.rust-lang.org/book/ch11-03-test-organization.html#submodules-in-integration-tests when making this change.

File: iml-system-docker-tests/tests/ldiskfs_test.rs (hunk 1 above)
Comment (johnsonw) [reply iml-system-docker-tests/tests/ldiskfs_test.rs:21]: Doesn't look like it's complaining now. I'll remove.

File: iml-system-test-utils/src/lib.rs [hunk 5]
```diff
@@ -126,3 +129,35 @@ pub fn get_local_server_names<'a>(servers: &'a [&'a str]) -> Vec<String> {
         .map(move |x| format!("{}.local", x))
         .collect()
 }
+
+#[async_trait]
+pub trait CmdErrSos {
+    async fn handle_test_result(self, hosts: &[&str], prefix: &str) -> Result<(), CmdError>;
```
Comment (jgrund) [inline iml-system-test-utils/src/lib.rs:135]: Can we use generics here for the err type?

Then we wouldn't need two separate traits / impls.

Comment (jgrund) [reply iml-system-test-utils/src/lib.rs]: Fixed.

File: iml-system-docker-tests/tests/ldiskfs_test.rs (hunk 1 above)
Comment (jgrund) [reply iml-system-docker-tests/tests/ldiskfs_test.rs:21]: Fixed.

**Pull request merged**